pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,264 @@
+func (k *KubeResourceQuota) getOrCreateDelegate(clusterName logicalcluster.Name) (*resourcequota.QuotaAdmission, error) {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	delegate := k.delegates[clusterName]
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	// TODO cherry pick https://github.com/kubernetes/kubernetes/pull/110246 so we can get the right stop channel wired.
+	// Or maybe we need to watch ClusterWorkspaces and stop when they're deleted?
+	delegate, err := resourcequota.NewResourceQuota(k.userSuppliedConfiguration, 5, make(chan struct{}))
+	if err != nil {
+		return nil, err
+	}
+
+	delegate.SetExternalKubeInformerFactory(&SingleClusterKubeSharedInformerFactoryForResourceQuotas{
+		SharedInformerFactory: k.kubeInformers,
+		clusterName:           clusterName,
+	})
+
+	delegate.SetExternalKubeClientSet(k.kubeClusterClient.Cluster(clusterName))
+
+	delegate.SetQuotaConfiguration(k.quotaConfiguration)
+
+	if err := delegate.ValidateInitialization(); err != nil {
+		return nil, err
+	}
+
+	k.delegates[clusterName] = delegate
+
+	return delegate, nil
+}
+
+func (k *KubeResourceQuota) SetKubeClusterClient(kubeClusterClient *kubernetes.Cluster) {
+	k.kubeClusterClient = kubeClusterClient
+}
+
+func (k *KubeResourceQuota) SetKcpInformers(informers kcpinformers.SharedInformerFactory) {
+	k.clusterWorkspaceLister = informers.Tenancy().V1alpha1().ClusterWorkspaces().Lister()
+}
+
+func (k *KubeResourceQuota) SetKubeInformers(informers informers.SharedInformerFactory) {
+	k.kubeInformers = informers
+
+	indexers := cache.Indexers{
+		kubequotacontroller.IndexByLogicalCluster: func(obj interface{}) ([]string, error) {
+			q, ok := obj.(*corev1.ResourceQuota)
+			if !ok {
+				return []string{}, fmt.Errorf("%T was not a ResourceQuota", obj)
+			}
+
+			return []string{logicalcluster.From(q).String()}, nil
+		},
+		kubequotacontroller.IndexByLogicalClusterAndNamespace: func(obj interface{}) ([]string, error) {
+			q, ok := obj.(*corev1.ResourceQuota)
+			if !ok {
+				return []string{}, fmt.Errorf("%T was not a ResourceQuota", obj)
+			}
+
+			return []string{clusters.ToClusterAwareKey(logicalcluster.From(q), q.Namespace)}, nil
+		},
+	}
+
+	resourceQuotaIndexer := k.kubeInformers.Core().V1().ResourceQuotas().Informer().GetIndexer()
+
+	for indexName := range resourceQuotaIndexer.GetIndexers() {
+		delete(indexers, indexName)
+	}
+
+	if len(indexers) > 0 {
+		if err := resourceQuotaIndexer.AddIndexers(indexers); err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (k *KubeResourceQuota) SetQuotaConfiguration(quotaConfiguration quota.Configuration) {
+	k.quotaConfiguration = quotaConfiguration
+}
+
+// SingleClusterKubeSharedInformerFactoryForResourceQuotas embeds an informers.SharedInformerFactory instance, overriding
+// .Core().V1().ResourceQuotas() to be scoped to a single logical cluster. All other informers are untouched.
+type SingleClusterKubeSharedInformerFactoryForResourceQuotas struct {
```

Comment:
why public?

ScopedResourceQuotaInformerFactory?

Reply:
Fixed in squash commit. Unexported this as well as the 2 types below.